common/src/java/org/apache/hadoop/hive/conf/HiveConf.java

```diff
@@ -3710,7 +3710,12 @@ public static enum ConfVars {
     HIVE_EXPLAIN_USER("hive.explain.user", true,
         "Whether to show explain result at user level.\n" +
         "When enabled, will log EXPLAIN output for the query at user level. Tez only."),
-
+    HIVE_EXPLAIN_VISIT_LIMIT("hive.explain.visit.limit", 256, new RangeValidator(1, Integer.MAX_VALUE),
```

Comment:
Does this limit actually only apply when doing an EXPLAIN ?  Even if the query is not explicitly doing an explain, the limit should still apply during compilation right ?  Also, the explain plan is written to the log file so we would want to avoid writing huge plans there.  It seems to me a better config would be 'hive.compile.node.visit.limit'.

Reply:
The limit applies only when doing EXPLAIN thus the choice of this name. Adding `node` in the property name is a good idea so I applied this change (https://github.com/apache/hive/pull/3852/commits/5c9933e1a59fe6beeee83638cfa62f9ce887c711). I opted to introduce a limit cause it is not possible to address the problem at the EXPLAIN level without changing the output format.

There are many places where a graph is traversed in Hive and applying a global limit everywhere would be difficult to enforce. Moreover, it would possibly require changes in many places leading to a change with much bigger impact. 

If we want to go for a global visit limit then maybe it would be better to do it as a separate JIRA/PR.

Reply:
Thanks for the explanation. For now, I am good to keep this as is for EXPLAIN.  A global limit can be considered in the future but it's ok to wait for a real use case where that is needed.